File: back-end/worker-mngmt/Services/OfferService.cs
using Microsoft.EntityFrameworkCore;
using Microsoft.Extensions.Hosting;
using System;
using System.Collections;
using System.Collections.Generic;
using System.Linq;
using worker_mngmt.Data;
using worker_mngmt.DTOs;
using worker_mngmt.Models;

namespace worker_mngmt.Services
{
    public class OfferService : IOfferService
    {
        private AppDbContext _appDbContext;

        public OfferService(AppDbContext appDbContext)
        {
            _appDbContext = appDbContext;
        }

        public int AddReview(AddReviewDto addReviewDto)
        {
            var offer = _appDbContext.Offers.SingleOrDefault(o => o.Id.Equals(new Guid(addReviewDto.OfferId)));
            var review = offer.Reviews.SingleOrDefault(r => r.Reviewer.Id.Equals(new Guid(addReviewDto.ReviewerId)));
            var revewer = _appDbContext.Users.SingleOrDefault(u => u.Id.Equals(new Guid(addReviewDto.ReviewerId)));

            if(review == null)
            {
                _appDbContext.Reviews.Add( new Review()
                {
                    Id = Guid.NewGuid(),
                    Offer = offer,
                    Reviewer = revewer
                });
            }
            else
            {
                _appDbContext.Reviews.Remove(review);
            }

            _appDbContext.SaveChanges();

            return offer.Reviews.Count();
        }

    public void Apply(ApplyToOfferDto applyToOfferDto)
        {
            var offer = _appDbContext.Offers.FirstOrDefault(o => o.Id.Equals(new Guid(applyToOfferDto.OfferId)));
            var user = _appDbContext.Users.FirstOrDefault(u => u.Id.Equals(new Guid(applyToOfferDto.UserId)));
            
            _appDbContext.Candidates.Add(new Candidate()
            {
                Id = Guid.NewGuid(),
                Status = CandidateStatus.Pending,
                Offer = offer,
                User = user
            });

            _appDbContext.SaveChanges();
        }

        public IEnumerable<OfferDetailDto> GetAll()
        {
            return _appDbContext.Offers.Select(o => new OfferDetailDto()
            {
                Id = o.Id,
                Title = o.Ttile,
                Description = o.Description,
                Category = o.Category,
                Type = o.Type,
                CandidatesId = o.Candidates.Select(c => c.User.Id),
                Likes = o.Reviews.Count(),
                OrgId = o.Publisher.Id
            });
        }

        public IEnumerable<OfferApplicantDto> GetApplicants(string id)
        {
            return _appDbContext.Candidates.Where(c => c.Offer.Id.Equals(new Guid(id))).Select(c => new OfferApplicantDto()
            {
                Id = c.Id,
                Email = c.User.Email,
                Name = c.User.Name,
                Status = c.Status
            });
        }

        public OfferDetailDto GetById(string id)
        {
            var offer = _appDbContext.Offers.FirstOrDefault(o => o.Id.Equals(new Guid(id)));

            return new OfferDetailDto()
            {
                Id = offer.Id,
                Title = offer.Ttile,
                Description = offer.Description,
                Category = offer.Category,
                Type = offer.Type,
                Likes = offer.Reviews.Count(),
                OrgId = offer.Publisher.Id,
                CandidatesId = offer.Candidates.Select(c => c.User.Id)
            };
        }

        public IEnumerable<OfferDetailDto> GetOffers(string id)
        {
            return _appDbContext.Offers.Where(o => o.Publisher.Id.Equals(new Guid(id))).Select(o => new OfferDetailDto()
            {
                Id = o.Id,
                Title = o.Ttile,
                Description = o.Description,
                Category = o.Category,
                Type = o.Type,
                CandidatesId = o.Candidates.Select(c => c.User.Id),
                Likes = o.Reviews.Count(),
                OrgId = o.Publisher.Id
            });
        }

        public void GiveApplicantResponse(OfferResponseDto offerResponseDto)
        {
            var candidate = _appDbContext.Candidates.FirstOrDefault(c => c.Id.Equals(new Guid(offerResponseDto.CandidateId)));

            candidate.Status = offerResponseDto.IsApproved ? CandidateStatus.Approved : CandidateStatus.Rejected;

            _appDbContext.SaveChanges();
        }

        public OfferDetailDto NewOffer(NewOfferDto newOffer)
        {
            var offer = new Offer()
            {
                Id = Guid.NewGuid(),
                Ttile = newOffer.Title,
                Description = newOffer.Description,
                Type = newOffer.Type,
                Category = newOffer.Category,
                Publisher = _appDbContext.Orgs.SingleOrDefault(o => o.Id.Equals(new Guid(newOffer.OrgId)))
            };

            _appDbContext.Offers.Add(offer);
            _appDbContext.SaveChanges();

            return new OfferDetailDto()
            {
                Id = offer.Id,
                Title = offer.Ttile,
                Description = offer.Description,
                Category = offer.Category,
                Type = offer.Type,
                Likes = offer.Reviews.Count(),
                OrgId = offer.Publisher.Id,
                CandidatesId = offer.Candidates.Select(c => c.User.Id)
            };
        }

    }
}

File: front-end/worker-mngmt/src/app/services/offer.service.ts
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { Observable, of } from 'rxjs';
import { applicant } from '../data/interfaces/applicant';
import { offer } from '../data/models/offer';

@Injectable({
  providedIn: 'root'
})
export class OfferService {

  constructor(private httpClient:HttpClient) {

   }

   public getById(id: string):Observable<offer> {
    let url = `http://localhost:5000/api/OfferController/GetById/${id}`;

    return this.httpClient.get<offer>(url);
   }

   public getApplicants(id: string | undefined) : Observable<applicant[]> {
    let url = `http://localhost:5000/api/OfferController/GetApplicants/${id}`;

    return this.httpClient.get<applicant[]>(url);
   }

   public getOffers(id: string | undefined) : Observable<offer[]> {
    let url = `http://localhost:5000/api/OfferController/GetOffers/${id}`;

    return this.httpClient.get<offer[]>(url);
   }

   public giveReview(isPositive: boolean, id: string | undefined, userId: string) : Observable<number> {
    let url = `http://localhost:5000/api/OfferController/AddReview`;

    let body = {"ReviewerId":userId, "OfferId":id, "IsPositive":isPositive };
    let headers= {'Content-Type': 'application/json'}

    return this.httpClient.post<number>(url,body, { headers: headers});
   }

   public getAll() : Observable<offer[]> {
    let url = `http://localhost:5000/api/OfferController/GetAll`;

    return this.httpClient.get<offer[]>(url);
   }

   public apply(id: string | undefined, userId: string) : Observable<void> {
    let url = `http://localhost:5000/api/OfferController/Apply`;

    let body = {"UserId":userId, "OfferId":id};
    let headers= {'Content-Type': 'application/json'}

    return this.httpClient.post<void>(url,body, { headers: headers});
   }

   public giveApplicantResponse(isApproved: boolean, candidateId: string, offerId: string | undefined): Observable<string> {
    let url = `http://localhost:5000/api/OfferController/GiveApplicantResponse`;

    let body = {"isApproved":isApproved, "OfferId":offerId, "CandidateId": candidateId};
    let headers= {'Content-Type': 'application/json'}

    return this.httpClient.post<string>(url,body, { headers: headers});
   }

   public newOffer(title: string, description: string, type: string, category: string, orgId: string) : Observable<offer> {
    let url = `http://localhost:5000/api/OfferController/NewOffer`;
    
    let body = {"Title": title, "Description": description, "Type": type, "Category": category, "OrgId": orgId };
    let headers= {'Content-Type': 'application/json'}

    return this.httpClient.post<offer>(url, body, { headers: headers});
   }
}
